### PR TITLE
Private Events

### PR DESCRIPTION
**Pull request by emmanuellekamwa**
In this project,
- I have mainly made use of associations which is the aim of the project
- I have also used validators
- Rubocop and linters to fix errors

### CODE REVIEW

**Review by SmithVyne — CHANGES_REQUESTED** [review]
### Changes Required 
Hello there, your project is great but one of the two linters in this pull request is failing. You need to fix all the lint errors and re-submit the project for another review.
Make these changes and request another code review. You should get it approved this time.

You can also reach out to me on slack, **@SmithVyne** if you have any other questions.

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUESTED 🔴 ♻️ 
Hello @emmanuellekamwa 👋🏾 👋🏾 , I like that you have implemented the previous TSE comment. Your UI is clean and easy to navigate. Keep it up 👏🏾 👏🏾 
However, let's make a few improvements

- [ ] As a user, when I try to attend events created by other users, I get the following error

image


##### README
##### Setup

- [ ] First clone the repository on your local machine -> Give the actual git clone command so a new user would find it easy to go do `git clone <>`

- [ ] Browse into the project's directory by typing cd plus the name of the project file -> You can simplify this by saying `cd Private-events`

- [ ] Run bundle install to install the necessary gems -> After this statement don't forget to add the command `yarn` because i could not run your project without it

- [ ] Set up the database using rails db: migrate -> Before a user can migrate the db, they need to create the db first so include `rails db: create`
- [ ] After the migration, add a command on how to start your rails application ie `rails server

That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, or just comment on this PR and tag my name and I will be glad to help_
Regards
Peris Ndanu

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  <ul class="nav nav-pills pull-right d-flex justify-content-end">
+    <li class="links"><%= nav_link "Home", root_path %></li>
+    <li class="links"><%= nav_link "Events", events_path %></li>
+    <li class="links"><%= nav_link "Users", users_path %></li>
+    <li class="links"><%= nav_link "About", about_path %></li>
+    <% if signed_in? %>
+      <li class="links"><%= link_to "#{current_user.name} (Sign Out)", signout_path, method: "delete" %></li>
+    <% else %>
+      <li class="links"><%= link_to "Sign In", signin_path %></li>
+    <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/_header.html.erb]: There should be absolutely NO logic in views

File: app/views/shared/_user_home.html.erb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+<% if @previous_events.any? %>
+<h2>You Attended:</h2>
+<%= render 'shared/table', object: @previous_events %>
+<% end %>
+
+<% if @user.events.any? %>
+<h2>Your Created Events</h2>
+<%= render 'shared/table', object: @user.events %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_user_home.html.erb:19]: There should be absolutely NO logic in views

File: app/views/static_pages/_user_home.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+<% end %>
+
+<% if @previous_events.any? %>
+<h2>Attended</h2>
+<%= render partial: 'shared/table', locals: {events:@previous_events} %>
+<% end %>
+
+<% if @user.events.any? %>
+<h2>My Events</h2>
+<%= render partial: 'shared/table', locals: {events:@user.events} %>
```
Comment (acushlakoncept) [inline app/views/static_pages/_user_home.html.erb:20]: There should be absolutely NO logic in views

File: app/views/static_pages/home.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+<% else %>
+
+  <div class="jumbotron">
+    <h1>Plan an Event</h1>
+    <p class="lead">Keep track of events with your friends.  This is a simple app to organize your event planning.
+       </p>
+    <p><%= link_to "Create an Event", signup_path, class: "btn btn-lg btn-success" %></p>
+  </div>
+
+<% end %>
```
Comment (acushlakoncept) [inline app/views/static_pages/home.html.erb:14]: There should be absolutely NO logic in views

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+<h2>Attended</h2>
+<%= render partial: 'shared/table', locals: {events:@previous_events} %>
+
+<% end %>
+
+<% if @user.events.any? %>
+<h2>Created by <%= @user.name%></h2>
+
+<%= render partial: 'shared/table', locals: {events:@user.events} %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb]: There should be absolutely NO logic in views

File: config/routes.rb [hunk 6]
```diff
@@ -1,3 +1,75 @@
...
+  #   end
+  #   resources :posts, concerns: :toggleable
+  #   resources :photos, concerns: :toggleable
+
+  # Example resource route within a namespace:
+  #   namespace :admin do
+  #     # Directs /admin/products/* to Admin::ProductsController
+  #     # (app/controllers/admin/products_controller.rb)
+  #     resources :products
+  #   end
```
Comment (acushlakoncept) [inline config/routes.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @emmanuellekamwa,  the UI of your project is really presentable, and congrats on the level of progress.

However, there are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.

- [ ] The **Attend** button for upcoming events works fine but that of the past event throws an error. I would suggest you remove the attend button from the past event, after all, nobody goes to a past event.
Screen Shot 2021-03-12 at 10 47 36 AM

- [ ] When a user is logged in, he/she should not be able to access the signup page again. The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

### Tests

* [ ] Your models should have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [ ] Your authentication should have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [ ] Your event managment features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
+class User < ApplicationRecord
+  validates :name, presence: true
+  validates :email, presence: true
+
+  has_many :events, foreign_key: :creator_id
+
+  has_many :invites, foreign_key: :invitee_id
+  # has_many :attended_events, :through => :invites
```
Comment (enaburekhan) [inline app/models/user.rb]: Remove these commented-out codes in line 8

File: app/views/layouts/_header.html.erb (hunk 1 above)
Comment (enaburekhan) [inline app/views/layouts/_header.html.erb]: As highlighted by the last reviewer, there should be absolutely no logic in view

File: app/views/shared/_user_home.html.erb (hunk 2 above)
Comment (enaburekhan) [inline app/views/shared/_user_home.html.erb:19]: Move the logic to a helper function.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi Alix!
Great job on this project so far. The UI is awesome:clap::clap:. However, we just need to make a few changes to get this project approved:muscle:.**

**Mandatory requirements**

- [ ] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). When I try to locate this page, I get an error. see [screen-shot](https://imgur.com/a/t57rbpL) for reference.

- [ ] Some tests are failing on rspec. Also, see [screen-shot](https://imgur.com/a/ywiuGL7) for reference. Please fix this.

- [ ] As highlighted by the last reviewer. There should be absolutely no logic in view. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check out this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) on how to use a rails helper. See [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright. Move all the logics in view to a helper

- [ ] Remove commented-out codes in user.rb in model. They make the codebase to look untidy and the codes difficult to read and understand.

Make the above-required changes and when you are ready, request another review.
Have any question/comment? please drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @emmanuellekamwa  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] When I try to go on user's show page the app is crashing
Screenshot from 2021-03-16 13-16-50
- [ ] There is no way of inviting attendees to the event
- [ ] Some tests cases are failing to run, please fix them out
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **NOTE**

I approved the project by mistake, please work on the requested changes and I kindly ask you to wait for the button to be turned from "Submit' to "Changes Required" in your view progress page
Once again, I'm sorry you'll have to wait until that mistake is being fixed 🙏🙏

<hr>

# **Changes Required 🛑♻**

<hr>

Hello there :)
thanks for your efforts and for making most of the changes requested, but you haven't implement the changes from the last review. Please make sure to do the following changes.

- [ ] A user should be able to attend an event. Currently, I cannot see a way to do that!
- [ ] Some test cases are still failing, please work on them.

<hr>

Kindly make the requested changes and ask for another review.
good luck and happy coding. 👨‍💻💻😊

**Review by mahmoud717 — APPROVED** [review]
### STATUS: APPROVED ✔️ ✔️ 

**hi guys**

**Great job with making all of the requirements. the project is approved, you can go ahead and merge this pull request**

**Pull request merged**